src/main/java/com/aws/greengrass/shadowmanager/PubSubIntegrator.java

```diff
@@ -0,0 +1,173 @@
+            this.pubSubClientWrapper.subscribe(this::handlePublishedMessage);
+        }
+    }
+
+    /**
+     * Unsubscribes to the shadow topic over local PubSub.
+     */
+    public void unsubscribe() {
+        if (this.subscribed.compareAndSet(true, false)) {
+            this.pubSubClientWrapper.unsubscribe(this::handlePublishedMessage);
+        }
+
+    }
+
+    /**
+     * Helper function to extract the thingName, shadowName and shadow operation from mqtt topic.
+     *
+     * @param topic PubSub topic on which the message was sent.
+     * @return ShadowRequest object with shadow details
+     */
+    ShadowRequest extractShadowFromTopic(String topic) {
+        final Matcher matcher = shadowPattern.matcher(topic);
+
+        if (matcher.find() && matcher.groupCount() == 4) {
+            String thingName = matcher.group(1);
+            String shadowName = matcher.group(3);
+            String operation = matcher.group(4);
+            return new ShadowRequest(thingName, shadowName, operation);
+        }
+        logger.atWarn()
+                .kv("topic", topic)
+                .log("Unable to parse shadow topic for thing name, shadow name and shadow operation");
+        throw new IllegalArgumentException("Unable to parse shadow topic for thing name shadow name "
+                + "and shadow operation");
+    }
+
+    /**
+     * Handle the new message that is published oveg local PubSub. Extract the necessary information from the shadow
+     * topic to accurately route the message to the appropriate handler.
+     * It will ignore the message if it is unable to extract all the necessary information from the shadow topic.
+     *
+     * @param publishEvent The message that is published over local PubSub.
+     */
+    private void handlePublishedMessage(PublishEvent publishEvent) {
+        String topic = publishEvent.getTopic();
+        logger.atDebug().kv(LOG_TOPIC, topic).log("Processing new shadow operation message over local PubSub");
+        try {
+            ShadowRequest shadowRequest = extractShadowFromTopic(topic);
+            switch (shadowRequest.getOperation().toLowerCase()) {
+                case "update":
+                    UpdateThingShadowRequest request = new UpdateThingShadowRequest();
+                    request.setThingName(shadowRequest.getThingName());
+                    request.setShadowName(shadowRequest.getShadowName());
+                    request.setPayload(publishEvent.getPayload());
+
+                    this.updateThingShadowRequestHandler.handleRequest(request, SHADOW_MANAGER_NAME);
+                    logger.atTrace()
+                            .setEventType(LogEvents.UPDATE_THING_SHADOW.code())
+                            .kv(LOG_THING_NAME_KEY, shadowRequest.getThingName())
+                            .kv(LOG_SHADOW_NAME_KEY, shadowRequest.getShadowName())
+                            .log("Successfully updated shadow");
```

Comment:
Do you need these logs? The handlers themselves [log](https://github.com/aws-greengrass/aws-greengrass-shadow-manager/blob/main/src/main/java/com/aws/greengrass/shadowmanager/ipc/UpdateThingShadowRequestHandler.java#L227)

Reply:
True. I just thought it might be a good data point for the customer to know if this operation happened over IPC PubSub or IPC shadow. I can add a kv specific to pubsub to these logs. Hence I put it in trace too.

Reply:
but they would see the log on line 114 on the same thread that it is "Processing new shadow operation message over local pubsub" with the topic - which would include the operation right? 

It would look like:

```
[date] [DEBUG] (<thread-id-1>) c.a.g.sm.PubSubIntegrator: Processing new shadow operation message over local PubSub. {topic=$aws/things/<thingName>/shadow/<shadowName>/update}
[date] [INFO] (<thread-id-1>) c.a.g.sm.ipc.UpdateThingShadowRequestHandler: Successfully updated shadow. {thing name=<thingName>, shadow name=<shadowName>, local-version=<version>, service-name=<serviceName>}
[date] [TRACE] (<thread-id-1>) c.a.g.sm.PubSubIntegrator: handle-update-thing-shadow. Successfully updated shadow. {thing name=<thingName>, shadow name=<shadow Name>
```